coredns/plugin/handler.go

```diff
@@ -83,7 +83,10 @@ func (lh *Lighthouse) getDNSRecord(ctx context.Context, zone string, state *requ
 			pReq.service, lh.ClusterStatus.IsConnected)
 		if !found {
 			log.Debugf("No record found for %q", state.QName())
-			return lh.nextOrFailure(ctx, state.Name(), w, r, dns.RcodeNameError, "record not found")
+			if lh.Fall.Through(state.Name()) {
+				return plugin.NextOrFailure(lh.Name(), lh.Next, ctx, w, r) // nolint:wrapcheck // Let the caller wrap it.
+			}
+			return lh.emptyResponse(state, dns.RcodeNameError)
```

Comment:
In looking at the [coredns server code](https://github.com/coredns/coredns/blob/fc7f3835d506960364b21cf62116a70eb8f4caf2/core/dnsserver/server.go#L263):

```Go
rcode, _ := h.pluginChain.ServeDNS(ctx, w, r)
if !plugin.ClientWrite(rcode) {
    errorFunc(s.Addr, w, r, rcode)
}
return
```

it assumes `ServeDNS` has written the response body unless the returned code indicates it hasn't been written as determined by `ClientWrite` (which is only the codes `RcodeServerFailure`, `RcodeRefused`, `RcodeFormatError` and `RcodeNotImplemented`), in which case it writes the response. 

So  in `lh.nextOrFailure`:

```Go
func (lh *Lighthouse) nextOrFailure(ctx context.Context, name string, w dns.ResponseWriter, r *dns.Msg, code int, err string) (int, error) {
    if lh.Fall.Through(name) {
        return plugin.NextOrFailure(lh.Name(), lh.Next, ctx, w, r) // nolint:wrapcheck // Let the caller wrap it.
    }

    return code, lh.error(err)
}
```

if there's no next plugin, it simply returns the code but if it isn't one of the `ClientWrite` codes, then no response is written, which is the case here with `RcodeNameError`. 

Rather than using `emptyResponse` here, I think we should modify `nextOrFailure` to return a response similarly. To reduce code duplication, we can define a general function:

```Go
func (lh *Lighthouse) writeResponse(state *request.Request, a *dns.Msg) (int, error) {
    a.Authoritative = true

    wErr := state.W.WriteMsg(a)
    if wErr != nil {
        log.Errorf("Failed to write message %#v: %v", a, wErr)
	return dns.RcodeServerFailure, lh.error("failed to write response")
    }

    return dns.RcodeSuccess, nil
}
```

Then `emptyResponse` becomes:

```Go
func (lh *Lighthouse) emptyResponse(state *request.Request) (int, error) {
    a := new(dns.Msg)
    a.SetReply(state.Req)

    return lh.writeResponse(state, a)
}
```

And `nextOrFailure` becomes:

```Go
func (lh *Lighthouse) nextOrFailure(ctx context.Context, name string, w dns.ResponseWriter, r *dns.Msg, code int) (int, error) {
    if lh.Fall.Through(name) {
        return plugin.NextOrFailure(lh.Name(), lh.Next, ctx, w, r) // nolint:wrapcheck // Let the caller wrap it.
    }

    a := new(dns.Msg)
    a.SetRcode(state.Req, rcode)

    return lh.writeResponse(state, a)
}
```

Reply:
good idea

Reply:
[in coredns code ](https://github.com/coredns/coredns/blob/master/plugin/plugin.go#L20) 
```go
	// Handler is like dns.Handler except ServeDNS may return an rcode
	// and/or error.
	//
	// If ServeDNS writes to the response body, it should return a status
	// code. CoreDNS assumes *no* reply has yet been written if the status
	// code is one of the following:
	//
	// * SERVFAIL (dns.RcodeServerFailure)
	//
	// * REFUSED (dns.RecodeRefused)
	//
	// * FORMERR (dns.RcodeFormatError)
	//
	// * NOTIMP (dns.RcodeNotImplemented)
```

so only this 4 code can not use writeResponse method

Reply:
No it doesn't mean we can't handle those codes - it just means we don't have to. But we might as well to simplify our code.